static/js/lib/coupon.js

```diff
@@ -0,0 +1,91 @@
+// @flow
+import {
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_CONTENT_TYPE_COURSE,
+  COUPON_CONTENT_TYPE_COURSERUN,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../constants';
+import type {
+  Coupons,
+  Coupon,
+  CalculatedPrices,
+} from '../flow/couponTypes';
+import type {
+  CoursePrice,
+  CoursePrices,
+  Dashboard,
+} from '../flow/dashboardTypes';
+
+export const calculatePrices = (programs: Dashboard, prices: CoursePrices, coupons: Coupons): CalculatedPrices => {
+  type HasProgramId = {
+    program_id: number
+  };
+  function makeLookup<T: HasProgramId>(obj: Array<T>): Map<number, T> {
+    return new Map(obj.map((value: T) => [value.program_id, value]));
+  }
```

Comment:
It's hard to do for the other three functions since they reference the lookup via closure. `calcPriceForRun` would need to take 5 arguments if it were outside the function